tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java

```diff
@@ -0,0 +1,268 @@
+
+		charset = getCharset(properties);
+		commaByte = ",".getBytes(charset);
+		bracketOpenByte = "[".getBytes(charset);
+		bracketCloseByte = "]".getBytes(charset);
+
+		String fileName = getFileName(properties);
+		File file = new File(fileName).getAbsoluteFile();
+		file.getParentFile().mkdirs();
+
+		boolean append = Boolean.parseBoolean(properties.get("append"));
+		FileOutputStream stream = new FileOutputStream(file, append);
+
+		fileChannel = stream.getChannel();
+
+		writer = new OutputStreamWriter(stream);
+
+		boolean buffered = Boolean.parseBoolean(properties.get("buffered"));
+		if (buffered) {
+			writer = new BufferedWriterDecorator(writer);
+		}
+
+		boolean writingThread = Boolean.parseBoolean(properties.get("writingthread"));
+		if (!writingThread) {
+			writer = new SynchronizedWriterDecorator(writer, stream);
+		}
+		FileInputStream inputStream = null;
+		try {
+			inputStream = new FileInputStream(file);
+			inputChannel = inputStream.getChannel();
+			preprocessFile(append);
+		} finally {
+			if (inputStream != null) {
+				inputStream.close();
+			}
+		}
+	}
+
+	private Charset getCharset(final Map<String, String> properties) {
+		String charsetName = properties.get("charset");
+		try {
+			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
+		} catch (IllegalArgumentException ex) {
+			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
+			return Charset.defaultCharset();
+		}
+	}
+
+	private String getFileName(final Map<String, String> properties) {
+		String fileName = properties.get("file");
+		if (fileName == null) {
+			throw new IllegalArgumentException("File name is missing for file writer");
+		}
+		return fileName;
+	}
+
+	private void writeProperty(final Token token, final LogEntry logEntry, final StringBuilder builder,
+			final boolean hasPrevious) {
+		if (hasPrevious) {
+			builder.append(String.format(",%s", NEW_LINE));
+		}
+		token.render(logEntry, builder);
+	}
+
+	private String buildJsonLogEntry(final LogEntry logEntry, final StringBuilder builder) {
+		boolean hasMessage = logEntry.getMessage() != null;
+		boolean hasTimestamp = logEntry.getTimestamp() != null;
+		boolean hasLevel = logEntry.getLevel() != null;
+		boolean hasMethod = logEntry.getMethodName() != null;
+		boolean hasClass = logEntry.getClassName() != null;
+		boolean hasThread = logEntry.getThread() != null;
+
+		if (hasMessage) {
+			writeProperty(messageToken, logEntry, builder, false);
+		}
+		if (hasTimestamp) {
+			writeProperty(timestampToken, logEntry, builder, hasMessage);
+		}
+		if (hasLevel) {
+			writeProperty(levelToken, logEntry, builder, hasTimestamp);
+		}
+		if (hasClass) {
+			writeProperty(classToken, logEntry, builder, hasLevel);
+		}
+		if (hasMethod) {
+			writeProperty(methodToken, logEntry, builder, hasClass);
+		}
+		if (hasThread) {
+			writeProperty(threadToken, logEntry, builder, hasMethod);
+		}
+		return String.format(JSON_OBJECT, builder.toString());
```

Comment:
You could build the entire JSON object directly in the target string builder. It would result in an significant performance boost.

Reply:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998